internal/storage/image.go

```diff
@@ -398,6 +410,87 @@ func (svc *imageService) ImageStatus(systemContext *types.SystemContext, nameOrI
 	return &result, nil
 }
 
+// nolint:staticcheck
+func (svc *imageService) IsRunningImageAllowed(ctx context.Context, systemContext *types.SystemContext, configImage, userSpecifiedImage string) error {
+	if systemContext.SignaturePolicyPath == "" {
+		log.Infof(ctx, "No policy applied, skipping check if image is allowed to run")
+		return nil
+	} else if userSpecifiedImage == "" {
+		return errors.New("client did not set the CRI Config.Image.UserSpecifiedImage field, please update it to use signature policies")
+	}
+
+	log.Debugf(ctx, "Getting image status of config image: %s", configImage)
+	configImageStatus, err := svc.ImageStatus(systemContext, configImage)
+	if err != nil {
+		return fmt.Errorf("image status from config image: %w", err)
+	}
+
+	// Check the cache first
+	var policyContext *signature.PolicyContext
+	if cacheItem := svc.policyContextCache.Get(systemContext.SignaturePolicyPath); cacheItem != nil {
+		log.Debugf(ctx, "Using policy context from cache")
+		policyContext = cacheItem.Value()
+	} else {
+		log.Debugf(ctx, "Creating policy context")
+		policy, err := signature.DefaultPolicy(systemContext)
+		if err != nil {
+			return fmt.Errorf("get default policy: %w", err)
+		}
+
+		policyContext, err = signature.NewPolicyContext(policy)
+		if err != nil {
+			return fmt.Errorf("create policy context: %w", err)
+		}
+
+		log.Debugf(ctx, "Caching policy context")
+		svc.policyContextCache.Set(systemContext.SignaturePolicyPath, policyContext, ttlcache.DefaultTTL)
+	}
+
+	log.Debugf(ctx, "Parsing docker reference from %s", userSpecifiedImage)
+	imageRef, err := docker.ParseReference("//" + userSpecifiedImage)
+	if err != nil {
+		return err
+	}
+
+	log.Debugf(ctx, "Creating image source")
+	imageSrc, err := imageRef.NewImageSource(ctx, systemContext)
+	if err != nil {
+		var e errcode.Error
+		if as := errors.As(err, &e); as && e.ErrorCode().Descriptor().HTTPStatusCode == http.StatusUnauthorized {
+			return fmt.Errorf("verifying policies against private registries it not supported yet: %w", err)
+		}
+		return fmt.Errorf("create image source: %w", err)
+	}
+
+	// Connect the on-registry image with the already locally-pulled one
+	imageCloser, err := svc.PrepareImage(systemContext, userSpecifiedImage)
+	if err != nil {
+		return fmt.Errorf("prepare user specified image: %w", err)
+	}
+	defer imageCloser.Close()
+	if imageCloser.ConfigInfo().Digest.Encoded() != configImageStatus.ID {
```

Comment:
⚠️ 

Uh… comparing `ConfigInfo.Digest` values _could_ be sufficient to establish image identity in theory, but it isn’t right now; the way image pull works, nothing ever cares about `config.RootFS.DiffIDs`, so “the same config” does not imply “the same layers”.

That didn’t previously matter in the `PullImage` “a correctly-signed image already exists locally” fast path, because the signature policy in the original pull and on reuse would have been the same.[1]

That can’t be assumed now, and the registry could be colluding with one of the tenants. So an attacker tenant could trigger a pull of a malicious image with a config digest matching a victim image (with a permissive policy, or with a policy that requires a signature that exists for the malicious image), and then instruct the registry to serve the correctly-signed victim image. A victim tenant would then check the signature for the victim image, see that the config digest matches, and accept the already-pulled malicious image.

We would need to match the whole manifest bytes (or equivalently the manifest digest) from the signature-verified `UnparsedImage`, and match that against _one of_ the manifests stored with a `*storage.Image`. That sounds… possible but tricky and difficult.

[1](And it doesn’t matter in `podman pull` or CRI-O’s `ImagePull`, other than the “image exists” shortcut added in this PR, because on those paths we first deduplicate layers based on trusted data, and require layers to be equal while deduplicating by config digest.)

Reply:
IOW, _within local c/storage_, the same config digest implies the same image ID implies exactly the same image.

On registries on the internet, the same config digest (without comparing the rest of the manifest, and with a pull implementation that doesn’t care about config’s `RootFS.DiffIDs`) does not guarantee exactly the same image _if the two were pulled_.

Reply:
This code path does not exist any more.